File: backend-legacy/services/bedrock_service.py
import boto3
import json
import asyncio
import logging
from botocore.config import Config
from botocore.exceptions import ClientError
from typing import Optional, List, AsyncGenerator, Dict
from config.config import settings

logger = logging.getLogger(__name__)

class BedrockService:
    def __init__(self):
        # Recommended configuration for Bedrock streaming and timeouts
        config = Config(
            read_timeout=900,
            retries={'max_attempts': 3, 'mode': 'standard'}
        )
        self.client = boto3.client(
            service_name='bedrock-runtime',
            region_name=settings.AWS_REGION,
            aws_access_key_id=settings.AWS_ACCESS_KEY_ID,
            aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY,
            config=config
        )

    async def _invoke_model_async(self, model_id: str, body: dict, operation_name: str):
        """Helper to run synchronous boto3 call in a thread."""
        loop = asyncio.get_running_loop()
        request_body = json.dumps(body)
        try:
            return await loop.run_in_executor(
                None,
                lambda: self.client.invoke_model(
                    modelId=model_id,
                    body=request_body,
                    accept="application/json",
                    contentType="application/json"
                )
            )
        except ClientError as e:
            logger.error(f"Error during Bedrock {operation_name} for model {model_id}: {e}")
            raise  # Re-raise the exception to be handled by the caller

    async def create_text_embedding(self, text: str) -> List[float]:
        """Creates an embedding for a given text using Titan Text Embeddings."""
        model_id = "amazon.titan-embed-text-v1"
        body = {"inputText": text}
        try:
            response = await self._invoke_model_async(model_id, body, "text embedding")
            response_body = json.loads(response['body'].read())
            return response_body['embedding']
        except Exception as e:
            logger.error(f"Failed to create text embedding: {e}")
            # Depending on desired behavior, you might return an empty list or re-raise
            return []

    async def create_batch_text_embeddings(self, texts: List[str]) -> List[List[float]]:
        """Creates embeddings for a batch of texts concurrently."""
        tasks = [self.create_text_embedding(text) for text in texts]
        embeddings = await asyncio.gather(*tasks)
        return embeddings

    async def create_image_embedding(self, image_base64: str, text: Optional[str] = None) -> List[float]:
        """Creates an embedding for a given image (and optional text) using Titan Multimodal Embeddings."""
        model_id = "amazon.titan-embed-image-v1"
        body = {"inputImage": image_base64}
        if text:
            body["inputText"] = text
        try:
            response = await self._invoke_model_async(model_id, body, "image embedding")
            response_body = json.loads(response['body'].read())
            return response_body['embedding']
        except Exception as e:
            logger.error(f"Failed to create image embedding: {e}")
            return []

    async def generate_image(self, prompt: str) -> str:
        """Generates an image from a text prompt using Titan Image Generator."""
        model_id = "amazon.titan-image-generator-v1"
        body = {
            "taskType": "TEXT_IMAGE",
            "textToImageParams": {"text": prompt},
            "imageGenerationConfig": {
                "numberOfImages": 1,
                "quality": "standard",
                "height": 1024,
                "width": 1024,
                "cfgScale": 8.0,
                "seed": 0 # Using a fixed seed for reproducibility, can be randomized
            }
        }
        try:
            response = await self._invoke_model_async(model_id, body, "image generation")
            response_body = json.loads(response.get('body').read())
            return response_body.get('images')[0]
        except Exception as e:
            logger.error(f"Failed to generate image: {e}")
            return ""


    async def converse_stream(
        self,
        model_id: str,
        messages: List[Dict],
        system_prompt: str,
        tool_config: Optional[Dict] = None,
        temperature: Optional[float] = None,
        top_p: Optional[float] = None
    ) -> AsyncGenerator[Dict, None]:
        """
        Invokes a model with a streaming response using the Converse API.
        """
        loop = asyncio.get_running_loop()
        if not model_id:
            model_id = "anthropic.claude-3-5-sonnet-20240620-v1:0"

        inference_config = {}
        if temperature is not None:
            inference_config["temperature"] = temperature
        if top_p is not None:
            inference_config["topP"] = top_p
        
        # Converse API expects system prompt as a list of content blocks
        system_messages = [{"text": system_prompt}] if system_prompt else []
        
        try:
            response = await loop.run_in_executor(
                None,
                lambda: self.client.converse_stream(
                    modelId=model_id,
                    messages=messages,
                    system=system_messages,
                    toolConfig=tool_config,
                    inferenceConfig=inference_config
                )
            )
            
            stream = response.get('stream')
            if stream:
                for event in stream:
                    yield event
        except ClientError as e:
            logger.error(f"Error during Bedrock converse_stream for model {model_id}: {e}")
            # In a streaming context, you might want to yield an error event
            yield {"error": str(e)}
        except Exception as e:
            logger.error(f"An unexpected error occurred in converse_stream: {e}")
            yield {"error": str(e)}

bedrock_service = BedrockService() 

File: backend/src/services/bedrockService.ts
import { BedrockRuntimeClient, InvokeModelCommand } from '@aws-sdk/client-bedrock-runtime';
import { fromEnv } from '@aws-sdk/credential-provider-env';

export class BedrockService {
  private client: BedrockRuntimeClient;

  constructor() {
    this.client = new BedrockRuntimeClient({
      region: process.env.AWS_REGION || 'us-east-1',
    });
  }

  async createBatchTextEmbeddings(
    texts: string[], 
    onProgress?: (completed: number, total: number) => void
  ): Promise<number[][]> {
    if (!texts || texts.length === 0) {
      return [];
    }

    console.log(`🔮 Starting batch embedding for ${texts.length} texts`);
    
    try {
      // สำหรับข้อความจำนวนมาก ให้แบ่งเป็น batch ย่อย
      if (texts.length > 50) {
        return await this.createLargeBatchEmbeddings(texts, onProgress);
      }
      
      // สำหรับข้อความน้อย ใช้ parallel processing
      return await this.createParallelEmbeddings(texts, onProgress);
      
    } catch (error) {
      console.error('Error creating batch embeddings:', error);
      throw error;
    }
  }

  private async createLargeBatchEmbeddings(
    texts: string[], 
    onProgress?: (completed: number, total: number) => void
  ): Promise<number[][]> {
    const BATCH_SIZE = 10; // จำนวน embedding ที่ทำพร้อมกัน
    const DELAY_BETWEEN_BATCHES = 200; // หน่วงเวลาระหว่าง batch (ms)
    
    const allEmbeddings: number[][] = [];
    let completed = 0;
    
    console.log(`📦 Processing ${texts.length} texts in batches of ${BATCH_SIZE}`);
    
    for (let i = 0; i < texts.length; i += BATCH_SIZE) {
      const batch = texts.slice(i, i + BATCH_SIZE);
      
      try {
        console.log(`🔄 Processing batch ${Math.floor(i / BATCH_SIZE) + 1}/${Math.ceil(texts.length / BATCH_SIZE)}`);
        
        const batchEmbeddings = await this.createParallelEmbeddings(batch);
        allEmbeddings.push(...batchEmbeddings);
        
        completed += batch.length;
        onProgress?.(completed, texts.length);
        
        // หน่วงเวลาเล็กน้อยเพื่อป้องกัน rate limiting
        if (i + BATCH_SIZE < texts.length) {
          await new Promise(resolve => setTimeout(resolve, DELAY_BETWEEN_BATCHES));
        }
        
      } catch (error) {
        console.error(`❌ Error in batch ${Math.floor(i / BATCH_SIZE) + 1}:`, error);
        
        // ลองทำทีละอันสำหรับ batch ที่ล้มเหลว
        console.log(`🔄 Retrying batch ${Math.floor(i / BATCH_SIZE) + 1} sequentially`);
        const fallbackEmbeddings = await this.createSequentialEmbeddings(batch);
        allEmbeddings.push(...fallbackEmbeddings);
        
        completed += batch.length;
        onProgress?.(completed, texts.length);
      }
    }
    
    console.log(`✅ Completed batch embedding: ${allEmbeddings.length} embeddings`);
    return allEmbeddings;
  }

  private async createParallelEmbeddings(
    texts: string[], 
    onProgress?: (completed: number, total: number) => void
  ): Promise<number[][]> {
    const MAX_CONCURRENT = 5; // จำกัดจำนวน concurrent requests
    const embeddings: number[][] = [];
    let completed = 0;
    
    // แบ่งเป็น chunks เพื่อจำกัด concurrent requests
    for (let i = 0; i < texts.length; i += MAX_CONCURRENT) {
      const chunk = texts.slice(i, i + MAX_CONCURRENT);
      
      const chunkPromises = chunk.map(async (text, index) => {
        try {
          const embedding = await this.createTextEmbedding(text);
          return { index: i + index, embedding, success: true };
        } catch (error) {
          console.warn(`⚠️ Failed to embed text ${i + index}, will retry`);
          return { index: i + index, embedding: null, success: false, error };
        }
      });
      
      const chunkResults = await Promise.allSettled(chunkPromises);
      
      // Process results and handle failures
      for (const result of chunkResults) {
        if (result.status === 'fulfilled') {
          const { index, embedding, success } = result.value;
          
          if (success && embedding) {
            embeddings[index] = embedding;
          } else {
            // Retry failed embedding sequentially
            try {
              console.log(`🔄 Retrying embedding for index ${index}`);
              const retryEmbedding = await this.createTextEmbedding(texts[index]);
              embeddings[index] = retryEmbedding;
            } catch (retryError) {
              console.error(`❌ Final failure for text ${index}:`, retryError);
              throw new Error(`Failed to create embedding for text at index ${index}`);
            }
          }
        } else {
          throw new Error(`Chunk processing failed: ${result.reason}`);
        }
        
        completed++;
        onProgress?.(completed, texts.length);
      }
    }
    
    return embeddings.filter(e => e); // Remove any null/undefined entries
  }

  private async createSequentialEmbeddings(texts: string[]): Promise<number[][]> {
    const embeddings: number[][] = [];
    
    for (let i = 0; i < texts.length; i++) {
      try {
        console.log(`🔄 Sequential embedding ${i + 1}/${texts.length}`);
        const embedding = await this.createTextEmbedding(texts[i]);
        embeddings.push(embedding);
        
        // Small delay to prevent overwhelming the service
        if (i < texts.length - 1) {
          await new Promise(resolve => setTimeout(resolve, 100));
        }
      } catch (error) {
        console.error(`❌ Failed to create embedding ${i + 1}:`, error);
        throw error;
      }
    }
    
    return embeddings;
  }

  async createTextEmbedding(text: string): Promise<number[]> {
    try {
      console.log('🔮 Creating embedding for text:', text.substring(0, 100) + '...');
      
      // Validate input
      if (!text || text.trim().length === 0) {
        throw new Error('Empty text provided for embedding');
      }
      
      // Truncate text if too long (Titan limit is ~8000 tokens)
      const maxLength = 25000; // Conservative estimate for characters
      const truncatedText = text.length > maxLength ? text.substring(0, maxLength) : text;
      
      if (text.length !== truncatedText.length) {
        console.warn(`⚠️ Text truncated from ${text.length} to ${truncatedText.length} characters`);
      }

      const input = {
        inputText: truncatedText,
      };

      const command = new InvokeModelCommand({
        modelId: 'amazon.titan-embed-text-v1',
        contentType: 'application/json',
        body: JSON.stringify(input),
      });

      const response = await this.client.send(command);
      
      if (!response.body) {
        throw new Error('Empty response from Bedrock');
      }
      
      const responseBody = JSON.parse(new TextDecoder().decode(response.body));
      
      if (!responseBody.embedding || !Array.isArray(responseBody.embedding)) {
        throw new Error('Invalid embedding response format');
      }
      
      console.log('✅ Embedding created successfully, dimension:', responseBody.embedding.length);
      return responseBody.embedding;
      
    } catch (error: any) {
      console.error('❌ Error creating text embedding:', {
        message: error instanceof Error ? error.message : error,
        name: error instanceof Error ? error.name : 'Unknown',
        code: error.code || 'Unknown',
        statusCode: error.$metadata?.httpStatusCode
      });
      
      // Check for specific AWS errors
      if (error.name === 'ThrottlingException') {
        throw new Error('Bedrock service is throttled. Please try again later.');
      }
      if (error.name === 'ValidationException') {
        throw new Error('Invalid input for embedding model.');
      }
      if (error.name === 'AccessDeniedException') {
        throw new Error('Access denied to Bedrock service. Check AWS credentials.');
      }
      if (error.name === 'ServiceUnavailableException') {
        throw new Error('Bedrock service is temporarily unavailable.');
      }
      
      // For development/testing, you might want to return dummy embeddings
      // Comment out the lines below for production
      if (process.env.NODE_ENV === 'development') {
        console.log('⚠️ Development mode: Returning dummy embedding with 1536 dimensions');
        return new Array(1536).fill(0.001); // Titan embedding dimension
      }
      
      // Re-throw the error in production
      throw error;
    }
  }

  async createImageEmbedding(imageBase64: string, text?: string): Promise<number[]> {
    const modelId = 'amazon.titan-embed-image-v1';
    const body: any = { inputImage: imageBase64 };
    if (text) body.inputText = text;
    try {
      const command = new InvokeModelCommand({
        modelId,
        body: JSON.stringify(body),
        contentType: 'application/json',
        accept: 'application/json',
      });
      const response = await this.client.send(command);
      const responseBody = JSON.parse(new TextDecoder().decode(response.body));
      return responseBody.embedding;
    } catch (error) {
      console.error('Failed to create image embedding:', error);
      return [];
    }
  }

  async generateImage(prompt: string): Promise<string> {
    const modelId = 'amazon.titan-image-generator-v1';
    const body = {
      taskType: 'TEXT_IMAGE',
      textToImageParams: { text: prompt },
      imageGenerationConfig: {
        numberOfImages: 1,
        quality: 'standard',
        height: 1024,
        width: 1024,
        cfgScale: 8.0,
        seed: 0,
      },
    };
    try {
      const command = new InvokeModelCommand({
        modelId,
        body: JSON.stringify(body),
        contentType: 'application/json',
        accept: 'application/json',
      });
      const response = await this.client.send(command);
      const responseBody = JSON.parse(new TextDecoder().decode(response.body));
      return responseBody.images?.[0] || '';
    } catch (error) {
      console.error('Failed to generate image:', error);
      return '';
    }
  }

  async *converseStream(
    modelId: string,
    messages: any[],
    systemPrompt: string,
    toolConfig?: any,
    temperature?: number,
    topP?: number
  ): AsyncGenerator<any, void, unknown> {
    // หมายเหตุ converseStream ยังไม่มีใน JS SDK v3 (2024/07) ใช้ invokeModel แบบปกติแทน
    // หรือถ้า Bedrock รองรับ streaming ในอนาคต ให้เปลี่ยนมาใช้ method ที่เหมาะสม
    const body: any = {
      messages,
      system: systemPrompt ? [{ text: systemPrompt }] : [],
      toolConfig,
      inferenceConfig: {},
    };
    if (temperature !== undefined) body.inferenceConfig.temperature = temperature;
    if (topP !== undefined) body.inferenceConfig.topP = topP;
    try {
      const command = new InvokeModelCommand({
        modelId,
        body: JSON.stringify(body),
        contentType: 'application/json',
        accept: 'application/json',
      });
      const response = await this.client.send(command);
      const responseBody = JSON.parse(new TextDecoder().decode(response.body));
      // จำลอง streaming: yield ทั้งหมดทีเดียว (ถ้าอนาคต SDK รองรับ streaming ให้แก้ไขตรงนี้)
      yield responseBody;
    } catch (error) {
      console.error('Error during converseStream:', error);
      yield { error: error instanceof Error ? error.message : 'Unknown error' };
    }
  }
}

export const bedrockService = new BedrockService(); 